[parse] Parse.Object.toJSON and some generic static methods

rdhelms opened the pull request:
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <http://parseplatform.org/Parse-SDK-JS/api/2.10.0/>
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

-----

Up until now, `Parse.Object.toJSON` has been typed as `any`, which ultimately defeats the purpose of having a typed Parse Object at all. This PR types the return of `toJSON` in accordance with the Parse library's [ParseObject.toJSON](https://github.com/parse-community/Parse-SDK-JS/blob/master/src/ParseObject.js#L436) logic. The `ToJSON` type uses another new `Encode` type to map types in accordance with the library's [encode function](https://github.com/parse-community/Parse-SDK-JS/blob/master/src/encode.js#L22). 

In order for the type to reflect the recursive behavior of `encode` itself, this did require bumping the TS version to 3.7 - but we had already been bumped up to 3.5, so this isn't as large of a jump as we made a few weeks ago from 2.x to 3.x.

Additionally, some methods needed to be made generic in order to have better control over their return type.

LinusU commented on types/parse/index.d.ts:
```diff
@@ -828,15 +855,15 @@ namespace Parse {
-        logIn(username: string, password: string, options?: FullOptions): Promise<User>;
-        logOut(): Promise<User>;
-        requestPasswordReset(email: string, options?: SuccessFailureOptions): Promise<User>;
+        become<T extends User>(sessionToken: string, options?: UseMasterKeyOption): Promise<T>;
```
Would you mind explaining why you chose `<T extends User>(...): Promise<T>` over
`<T extends Attributes>(...): Promise<User<T>>`? Not saying that it's wrong, just curious...

rdhelms replied:
Yep, I actually initially typed it as `<T extends Attributes>` but I changed it so that users would be able to pass a type that extended User but was not User itself. In other words, if someone made a custom User class that extended Parse.User, then I wanted them to be able to have these methods return their custom class instead of the parent Parse.User class.

Open to suggestions, though, if you have any thoughts on that.

LinusU replied:
Ah, of course, I think that sounds very good 👍

RaschidJFR replied:
Hey this works great! Actually this has been always bugging me since forever. Tnx @rdhelms

rdhelms commented:
I just added a quick addition to this that we noticed today - `Parse.Object.saveAll` previously only returned array types, and so trying to destructure the response only ever gave you a union of the types of the elements of the array being passed...but using a tuple type gives better inferences when destructuring.

One of the test cases I just added demonstrates this:
```ts
// $ExpectType [Object<Attributes>, Object<{ example: string; }>]
await Parse.Object.saveAll<[ typeof objUntyped, typeof objTyped ]>([ objUntyped, objTyped ]);
```

Previously, the `saveAll` type parameter would not have allowed passing a tuple type, and the return would have been `(Object<Attributes> | Object<{ example: string; }>)[]`

RaschidJFR commented:
Hey guys, I'm doubting now that upgrading TS to v7 is a good idea. This is a **breaking change** that got through unnoticed.

For example. **Angular 8** [that do not support ts +v3.5](https://gist.github.com/LayZeeDK/c822cc812f75bb07b7c55d07ba2719b3), and v9 is not yet released. 

I personally work a lot on Angular and sub frameworks (such as Ionic). Shouldn't we try a workaround or maybe revert part of this PR until the environment is more mature?
